### PR TITLE
Mainnet Release v2025.0.0

### PR DESCRIPTION
**Pull request by sophoah**
#4762 - Improve P2P Stream Handling & Stream Sync Enhancements
#4772 - Enhance Concurrency and Lock Management in Staged Stream Sync Module
#4777 - Add Support for Trusted Nodes in P2P Host
#4786 - Enhance Final Commit Function with Improved Error Handling
#4791 - Improve logging for consensus leader setup
#4792 - [fix] - CI - update macos from 12 to 13, sync the CI steps for macos
#4787 - Optimize Syncing-Consensus Communication to Prevent Unnecessary Consensus Resets
#4790 - Moved the `UpdateConsensusInformation` method to the test suite
#4801 - Add Logging to ReadySignal Method for New Block Proposals
#4785 - Additional logs for consensus
#4793 - localnet with multi bls
#4798 - Enhance Leader Rotation Logic to Address Edge Cases in Leader Selection
#4805 - Add Check for Faucet Key Existence Before Building External Validator
#4807 - Add Localnet Config to Adjust Blocks per Epoch
#4803 - Minor: Moved method `WaitForConsensusReadyV2` from `Consensus` struct.
#4799 - Improve Consensus logic for Multi-BLS Validators with quorum
#4811 - Fix Multi-BLS Key Validator Leader Rotation to Prevent Multiple Block Proposals
#4804 - Prevents BlocksSynchronized being called before finalCommit
#4815 - Add Resource Manager to P2P Host
#4818 - Add New Command to Makefile for Enhanced Localnet Debugging
#4816 - Fix for excess propose signals.
#4819 - multi bls multi ext node
#4827 - Add Resource Manager to P2P Configs When Enabled
#4832 - refactor: CI - fallback to the master branch in harmony-test if same branch name doesn't exitst
#4767 - fix: fix slice init length
#4759 - Effective gas price
#4834 - update harmony and bootnode version

### CODE REVIEW

**Comment by Frozen** [comment]
becase we gave `NthNextValidatorV2` release requires hardfork

**Review by GheisMohammadi — APPROVED** [review]

File: api/service/stagedstreamsync/beacon_helper.go [hunk 1]
```diff
@@ -55,7 +58,15 @@ func (bh *beaconHelper) start() {
 }
 
 func (bh *beaconHelper) close() {
-	close(bh.closeC)
+	bh.lock.Lock()
```
Comment (Frozen) [inline api/service/stagedstreamsync/beacon_helper.go:61]: Channels by default thread safe, no reason having additional lock here

Comment (sophoah) [reply api/service/stagedstreamsync/beacon_helper.go:61]: Can we have separate PR to remove it?

File: api/service/stagedstreamsync/downloader.go [hunk 2]
```diff
@@ -119,8 +134,9 @@ func (d *Downloader) DownloadAsync() {
 	select {
 	case d.downloadC <- struct{}{}:
 		consensusTriggeredDownloadCounterVec.With(d.promLabels()).Inc()
-
-	case <-time.After(100 * time.Millisecond):
+	case <-d.closeC:
```
Comment (Frozen) [inline api/service/stagedstreamsync/downloader.go:137]: downloadC initialized with `make(chan struct{}, 1)`, so it could be send even closed

**Comment by Frozen** [comment]
https://github.com/harmony-one/harmony/pull/4831/files#diff-46a911a8e03d9a1646d25e0d42a0b26b073362154647a437053fe2ec22e10c55R61

Deadlock

**Comment by Frozen** [comment]
https://github.com/harmony-one/harmony/pull/4831/files#diff-46a911a8e03d9a1646d25e0d42a0b26b073362154647a437053fe2ec22e10c55R67
Deadlock

File: api/service/stagedstreamsync/beacon_helper.go (hunk 1 above)
Comment (Frozen) [reply api/service/stagedstreamsync/beacon_helper.go:61]: minor, can be fixed in next patch

File: api/service/stagedstreamsync/downloader.go (hunk 2 above)
Comment (Frozen) [reply api/service/stagedstreamsync/downloader.go:137]: seems like minor

**Comment by sophoah** [comment]
@Frozen thanks yeah we are not using streamsync in mainnet nor testnet right now. Appreciate your approval.

File: consensus/consensus_v2.go [hunk 3]
```diff
@@ -138,7 +138,21 @@ func (consensus *Consensus) HandleMessageUpdate(ctx context.Context, peer libp2p
 	return nil
 }
 
-func (consensus *Consensus) finalCommit(isLeader bool) {
+func (consensus *Consensus) finalCommit(waitTime time.Duration, viewID uint64, isLeader bool) {
+	consensus.getLogger().Info().Str("waitTime", waitTime.String()).
```
Comment (Frozen) [inline consensus/consensus_v2.go:142]: possible race condition, but fixed in https://github.com/harmony-one/harmony/pull/4825

Comment (Frozen) [reply consensus/consensus_v2.go:142]: minor

File: consensus/quorum/quorom_test.go [hunk 4]
```diff
@@ -549,32 +551,233 @@ func TestInvalidAggregateSig(test *testing.T) {
...
+	go func() {
+		c.NthNextValidatorV2(slots, &list[1], 1)
+
+		done <- true
+	}()
+
+	select {
+	case <-done:
+		t.Log("Test completed successfully ")
+	case <-time.After(5 * time.Second):
```
Comment (Frozen) [inline consensus/quorum/quorom_test.go:730]: minor, avoid timeouts in tests

**Review by Frozen — APPROVED** [review]
